lib/ansible/module_utils/common/timeout.py

```diff
@@ -0,0 +1,50 @@
+# Copyright (c) 2020, Matt Martz <matt@sivel.net>
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import signal
+from functools import wraps
+
+from ansible.module_utils.common.warnings import warn
+
+
+def _raise_timeout(signum, frame):
+    raise TimeoutError
+
+
+class Timeout:
+    def __init__(self, timeout, raising=False):
+        self._timeout = timeout
+        self._raising = raising
+        self.timed_out = False
+
+    def __call__(self, func):
+        @wraps(func)
+        def inner(*args, **kwargs):
+            func_name = getattr(func, '__qualname__', func.__name__)
+            signal.signal(signal.SIGALRM, _raise_timeout)
+            signal.alarm(self._timeout)
+            try:
+                return func(*args, **kwargs)
+            except TimeoutError:
+                if self._raising:
+                    raise
+                warn('Function call to %s timed out after %d seconds.' % (func_name, self._timeout))
+                return None
+            finally:
+                signal.signal(signal.SIGALRM, signal.SIG_IGN)
+        return inner
+
+    def __enter__(self):
```

Comment:
I think we should have an entry-check to protect from reuse:
```suggestion
    def __enter__(self):
        assert not self.timed_out
```